src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs

```diff
@@ -98,9 +98,12 @@ private void RegisterChangeTokenCallback(IChangeToken? token)
                 {
                     return;
                 }
```

Comment:
The change here is reasonable. I am wondering if it will be better if we can do the following instead?

```C#
            private void RegisterChangeTokenCallback(IChangeToken? token)
            {
                if (token is null)
                {
                    return;
                }

                if (token.HasChanged && token.ActiveChangeCallbacks)
                {
                     OnChangeTokenFired();
                }
                else
                {
                     IDisposable registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>?)s)!.OnChangeTokenFired(), this);
                     SetDisposable(registraton);
                }
            }
```

I didn't try this code yet, but this will avoid the whole registration and disposing process.

Reply:
> The key point to keep in mind is that no changes should occur after calling reg.Dispose().

Thanks for the clarification. I was not getting this point in the first place.

Reply:
@tarekgh 
Thanks for review.  It seems to me that in this way we can break backward compatibility. What if someone for some reason decided to provide custom implementation of the IChangeToken interface that keeps track of the moment of registration, does some side effect (logging) through RegisterChangeCallback. Perhaps this is unlikely, but still. I would like to hear your opinion on this matter

Reply:
Although this is very unlikely, I agree with you in regards ensure registration in all cases as we don't know what can happen if we skip it even when the token is cancelled.